### PR TITLE
python3Packages.torch: enable vulkan support

### LINKED ISSUE #226998: Package request: pytorch with vulkan support
**Project description**

I wanted to build pytorch with vulkan support but I can't seem to find an existing way to do it. There doesn't seem to be a vulkan option... 😕 

**Metadata**

See https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/torch/default.nix.

### PR DESCRIPTION
**Pull request by zeuner**
Enables building with Vulkan support. For now, this can be used by installing `python3Packages.torchWithVulkan`.

Fixes https://github.com/NixOS/nixpkgs/issues/226998

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by FliegendeWurst** [comment]
By the way, if you add your git email as a secondary email on Github, the commits will show up as yours.

File: pkgs/development/python-modules/torch/default.nix [hunk 1]
```diff
@@ -375,6 +380,8 @@ buildPythonPackage rec {
       (lib.cmakeFeature "CMAKE_CUDA_COMPILER_TOOLKIT_VERSION" cudaPackages.cudaVersion)
     ];
 
+  USE_VULKAN = setBool vulkanSupport;
```
[the remark this answers is not in the document]
Comment (zeuner) [reply pkgs/development/python-modules/torch/default.nix]: Good idea, thanks!

File: pkgs/development/python-modules/torch/default.nix [hunk 2]
```diff
@@ -695,3 +708,6 @@ buildPythonPackage rec {
     broken = builtins.any trivial.id (builtins.attrValues brokenConditions);
   };
 }
+// lib.optionalAttrs vulkanSupport {
+  VULKAN_SDK = shaderc.bin;
```
[the remark this answers is not in the document]
Comment (zeuner) [reply pkgs/development/python-modules/torch/default.nix]: Same here.

**Review by FliegendeWurst — APPROVED** [review]

**Comment by jljusten** [comment]
> For now, this can be used by installing `python3Packages.torchWithVulkan`.

I notice that arch has had pytorch vulkan enabled since 1.3.1 Is there a known issue with vulkan support, or just being careful with enabling it by default?

Ref: https://gitlab.archlinux.org/archlinux/packaging/packages/python-pytorch/-/commit/ca50d3f98963a6c704a43015d83b7ff64377c61e

**Pull request merged**